### PR TITLE
Activity Tag filter added in matrix component

### PR DESCRIPTION
**Pull request by ptechofficial**
In regards to Issue #212, This commit will add the tags view in the task description. It also has some groundwork for the filter.

P.S. This is still a work in progress.

### CODE REVIEW

**Comment by wurstbrot** [comment]
cool, thank you @ptechofficial

**Comment by ptechofficial** [comment]
I am adding another filter below the sub-dimension filter. This will work like the chip filter only, there will be a link between the two filters(sub-dimension and activity). For eg, if a sub-dimension is removed, all the activity tags will also be removed. 

Currently I am working on that functionality.

**Comment by ptechofficial** [comment]
Hi @wurstbrot, I have successfully added Activity Tag Filter in Matrix Component. 

You can close this Issue now.

**Comment by ptechofficial** [comment]
Hi @wurstbrot, I have successfully added Activity Tag Filter in Matrix Component. 

You can close this Issue now.

File: src/app/component/matrix/matrix.component.html [hunk 1]
```diff
@@ -28,6 +28,34 @@
       </mat-option>
     </mat-autocomplete>
   </mat-form-field>
+  <mat-form-field class="example-chip-list">
+    <mat-label>Task Filter</mat-label>
+    <mat-chip-list #chipListtags aria-label="Filter" color="primary">
```
Comment (wurstbrot) [inline src/app/component/matrix/matrix.component.html:33]: chipListTags?

File: src/app/component/matrix/matrix.component.html [hunk 2]
```diff
@@ -28,6 +28,34 @@
...
+        (removed)="removeTask(task)"
+        [selected]="true">
+        {{ task }}
+        <button matChipRemove>
+          <mat-icon>cancel</mat-icon>
+        </button>
+      </mat-chip>
+      <input
+        #tagInput
+        [formControl]="rowCtrltags"
```
Comment (wurstbrot) [inline src/app/component/matrix/matrix.component.html:46]: **rowCtrlTags**

File: src/app/component/matrix/matrix.component.ts [hunk 3]
```diff
@@ -20,6 +20,7 @@ export interface MatrixElement {
 })
 export class MatrixComponent implements OnInit {
   MATRIX_DATA: MatrixElement[] = [];
+  // tag_data: string[] = [];
```
Comment (wurstbrot) [inline src/app/component/matrix/matrix.component.ts:23]: ?

File: src/app/component/matrix/matrix.component.ts [hunk 4]
```diff
@@ -32,6 +33,7 @@ export class MatrixComponent implements OnInit {
   allRows: string[] = [];
   dataSource: any = new MatTableDataSource<MatrixElement>(this.MATRIX_DATA);
   rowsCurrentlyBeingShown: string[] = [];
+  tasksCurrentlyBeingShown: string[] = [];
```
Comment (wurstbrot) [inline src/app/component/matrix/matrix.component.ts:36]: What is a task? Do you mean activity?

**Comment by wurstbrot** [comment]
Thank you, I like your PR.
I added some naming suggestions.

After removing a tag, how can I re-add it?

**Comment by ptechofficial** [comment]
I will update the name and changes you suggested. 

For re-adding a tag, there is an auto-complete function. But I just checked it and it isn't working properly. I'll fix it now.

**Review by 0x41head — CHANGES_REQUESTED** [review]
Hey @ptechofficial , thanks for the PR. I found a bug that is worth looking into.

**To reproduce the bug:**
- Select desired activity tags chips
- Remove the subdimension chip
- Add the same subdimension chip again

**Example:**
ezgif com-gif-maker


**Current result:**
All activities of sub-dimension displayed 

**Expected result:**
Same activity list as displayed previously.

I think a global variable which encapsulates the current matrix state would be sufficient.

Also, for design purposes a "no tags defined" chip would be great. 
@wurstbrot your thoughts ?

**Comment by ptechofficial** [comment]
Hi @0x41head, I have worked on the mentioned bug. 

Thanks for the suggestion, I have created a generalized function that will be called for both filters.
This will check both the tags ( SubDimension and Activity) and will present the result accordingly.

File: src/app/component/matrix/matrix.component.ts [hunk 5]
```diff
@@ -160,15 +160,15 @@ export class MatrixComponent implements OnInit {
...
+  filteredActivities: Observable<string[]>;
 
   autoCompeteResults: string[] = [];
   autoCompleteActivityResults: string[] = [];
 
   @ViewChild('rowInput') rowInput!: ElementRef<HTMLInputElement>;
   @ViewChild('tagInput') tagInput!: ElementRef<HTMLInputElement>;
 
-  updateActivity(tags: string[]): void {
+  updateActivity(): void {
```
Comment (0x41head) [inline src/app/component/matrix/matrix.component.ts:171]: can we change the function name to something like `updateActivitesBeingDisplayed` for clarity ?

File: src/app/component/matrix/matrix.component.ts [hunk 6]
```diff
@@ -130,26 +153,109 @@ export class MatrixComponent implements OnInit {
...
+          updatedActivities.push(temp);
+        }
+      }
+    }
 
-  //Remove chips
-  remove(row: string): void {
+    this.dataSource.data = JSON.parse(JSON.stringify(updatedActivities));
+  }
+  //Remove
```
Comment (wurstbrot) [inline src/app/component/matrix/matrix.component.ts:241]: how is this comment helpful?

File: src/app/component/matrix/matrix.component.ts [hunk 7]
```diff
@@ -130,26 +153,109 @@ export class MatrixComponent implements OnInit {
...
+  // Remove from SubDimension Filter
+  removeSubDimension(row: string): void {
```
Comment (wurstbrot) [inline src/app/component/matrix/matrix.component.ts:243]: Please remove the comment above and rename the function to:
removeFromSubDimensionFilter

Like that, you have the description in code and no comment is needed.

File: src/app/component/matrix/matrix.component.ts [hunk 8]
```diff
@@ -130,26 +153,109 @@ export class MatrixComponent implements OnInit {
...
     if (index >= 0) {
       this.rowsCurrentlyBeingShown.splice(index, 1);
     }
     this.autoCompeteResults.push(row);
-    this.dataSource.data.splice(index, 1);
-    this.dataSource._data.next(this.dataSource.data);
+    this.updateActivitesBeingDisplayed();
+  }
+  // Remove Task from Task Filter
+  removeActivity(activity: string): void {
```
Comment (wurstbrot) [inline src/app/component/matrix/matrix.component.ts:252]: Same here, rename to `removeAcitivityFromFilter` and remove the filter

File: src/app/component/matrix/matrix.component.ts (hunk 6 above)
Comment (ptechofficial) [reply src/app/component/matrix/matrix.component.ts]: I have removed it in the recent commit.

File: src/app/component/matrix/matrix.component.ts (hunk 7 above)
Comment (ptechofficial) [reply src/app/component/matrix/matrix.component.ts]: Okay, I have updated the names

**Review by wurstbrot — APPROVED** [review]

**Comment by wurstbrot** [comment]
> @wurstbrot what do you think about adding a chip called "tags not defined" for the activities that don't have tags ?

Good idea!

**Comment by ptechofficial** [comment]
Hi @0x41head 
Could you please explain this "tags not defined" feature?

**Comment by 0x41head** [comment]
We want to make sure that we include activities/tasks that do not have the attribute `tags`.
So we'd like to add a new chip called `tag not defined` to the filter tag.
This will filter out all actions with no `tags` attributes or with an empty `tags` property.

**Comment by ptechofficial** [comment]
Closing because of the new PR

**Pull request closed without merging**